[Tools] Add a library to build specialized generic metadata out of process.

mikeash opened the pull request:
This library uses GenericMetadataBuilder with a ReaderWriter that can read data and resolve pointers from MachO files, and emit a JSON representation of a dylib containing the built metadata.

We use LLVM's binary file readers to parse the MachO files and resolve fixups so we can follow pointers. This code is somewhat MachO specific, but could be generalized to other formats that LLVM supports.

rdar://116592577

mikeash committed:
[Tools] Add a library to build specialized generic metadata out of process.

This library uses GenericMetadataBuilder with a ReaderWriter that can read data and resolve pointers from MachO files, and emit a JSON representation of a dylib containing the built metadata.

We use LLVM's binary file readers to parse the MachO files and resolve fixups so we can follow pointers. This code is somewhat MachO specific, but could be generalized to other formats that LLVM supports.

rdar://116592577

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

al45tair commented on stdlib/public/ExternalGenericMetadataBuilder/ExternalGenericMetadataBuilder.cpp:
```diff
@@ -0,0 +1,2444 @@
+    if (command.C.cmd == llvm::MachO::LC_LOAD_DYLIB ||
+        command.C.cmd == llvm::MachO::LC_LOAD_WEAK_DYLIB ||
+        command.C.cmd == llvm::MachO::LC_REEXPORT_DYLIB ||
+        command.C.cmd == llvm::MachO::LC_LOAD_UPWARD_DYLIB) {
```
Do we need `LC_LAZY_LOAD_DYLIB`?

mikeash replied:
It seems not. `LC_LAZY_LOAD_DYLIB` seems to be obsolete, and this matches the behavior of dyld.

al45tair commented on test/ExternalGenericMetadataBuilder/ExternalGenericMetadataBuilder.swift:
```diff
@@ -0,0 +1,76 @@
+  let readJSONError = readJSONErrorCStr.map { String(cString: $0) }
+  expectNil(readJSONError)
+
+  // Find libswiftCore dynamically since we don't know where we might have loaded it from.
```
`swift_getRuntimeLibraryPath()` will tell you this, FWIW.

mikeash replied:
Thanks, that's much nicer.

mikeash commented on include/swift/Runtime/GenericMetadataBuilder.h:
```diff
@@ -488,57 +520,99 @@ class GenericMetadataBuilder {
-      case sizeWithAlignmentMask(2, 1, 0):
+      }
+      case sizeWithAlignmentMask(2, 1, 0): {
+        if (!VWT_Bi8_)
+          return *VWT_Bi8_.getError();
```
Ouch, major copy/pasta fail there. Good catch.

mikeash commented on stdlib/public/ExternalGenericMetadataBuilder/ExternalGenericMetadataBuilder.cpp:
```diff
@@ -0,0 +1,2444 @@
+      auto contents = section.getContents();
+      if (!contents) {
+        consumeError(contents.takeError());
+        continue;
```
That's a good idea. The MachO implementation can never return an error, but it'll be good in case of future expansion to other formats.

mikeash commented on stdlib/public/ExternalGenericMetadataBuilder/ExternalGenericMetadataBuilder.cpp:
```diff
@@ -0,0 +1,2444 @@
+        switch (cursor->getKind()) {
+        case Node::Kind::BoundGenericClass:
+        case Node::Kind::BoundGenericStructure:
+        case Node::Kind::BoundGenericEnum:
+        case Node::Kind::BoundGenericProtocol:
+        case Node::Kind::BoundGenericTypeAlias:
+        case Node::Kind::BoundGenericFunction:
+        case Node::Kind::BoundGenericOtherNominalType:
+          if (foundDepth == depth) {
+            auto typeList = cursor->getChild(1);
+            if (!typeList)
+              return BuilderError(
+                  "Requested generic parameter at depth=%" PRIu64
+                  " index=%" PRIu64
+                  ", BoundGeneric node does not have two children",
+                  depth, index);
+            if (typeList->getKind() != Node::Kind::TypeList)
+              return BuilderError(
+                  "Requested generic parameter at depth=%" PRIu64
+                  " index=%" PRIu64
+                  ", child 1 of BoundGeneric is %s, not TypeList",
+                  depth, index, nodeKindString(typeList->getKind()));
+
+            auto child = typeList->getChild(index);
+            if (!child)
+              return BuilderError(
+                  "Requested generic parameter at depth=%" PRIu64
+                  " index=%" PRIu64
+                  ", but bound generic TypeList only has %zu children",
+                  depth, index, typeList->getNumChildren());
+
+            return child;
+          }
+          foundDepth++;
+          break;
+
+        default:
+          break;
+        }
+      }
```
Oops. Looks like we don't yet have enough support for nested types for this to matter. I added code to walk the tree here, and we'll end up testing it as part of testing expanded support for nested types.

mikeash commented on stdlib/public/ExternalGenericMetadataBuilder/ExternalGenericMetadataBuilder.cpp:
```diff
@@ -0,0 +1,2444 @@
+ExternalGenericMetadataBuilderContext<Runtime>::addImagesInPath(
+    std::string path) {
+  auto forEachFile = [&](std::string filepath) {
+    auto error = addImageAtPath(filepath);
```
I'm not sure what the right approach is here. It doesn't matter _too_ much, as `swift_externalMetadataBuilder_addDylib` is intended to be the primary interface, and this is a convenience. For the convenience interface, I'd rather it try load everything in sight rather than have it potentially skip something we want to keep if somebody puts a weird extension on their dylib.

mikeash commented:
Addressed review feedback and moved the code to `lib` to join its library/tool friends. I also improved the test to be a bit more comprehensive and test the full JSON output rather than just looking for key strings.

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please clean test Windows platform

mikeash commented:
@swift-ci please test

mikeash commented:
@swift-ci please test macos platform
